### PR TITLE
HIP "Basic" Example Suite (part 3)

### PR DESCRIPTION
**Pull request by Snektron**
New examples:
* bandwidth: Program that measures memory bandwidth from host to device, device to host, and device to device.
* runtime compilation: Simple program that showcases how to use HIP runtime compilation (hipRTC) to compile a kernel and launch it on a device.
* shared memory: Showcases how to use static shared memory by implementing a simple matrix transpose kernel.
* 2d shuffle: Uses a simple matrix transpose kernel to showcase how to use warp shuffle operations.
* assembly to executable: Program and accompanying build systems that show how to manually compile and link a HIP application from host and device code.
* llvm ir to executable: Shows how to create a HIP executable from LLVM IR.

Fixes:
* Resolved a mistake that occured while rebasing the events example internally.
* Removed default target gfx90c from the saxpy Windows VS2019 files.
* Fixed some typos in earlier examples.
* Fixed VS2019 target names to start with the prefix of the library where they were missing.
* Fixed project names of some sln files, which were still set to those of the template.

Misc:
* Updated docker images to base upon ROCm 5.3 (and includes a small fix regarding a header file).
* Added required Vulkan packages to docker images (for the Vulkan example in the future).

### CODE REVIEW

File: HIP-Basic/assembly_to_executable/Makefile [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2020 - 2021 Advanced Micro Devices, Inc. All Rights Reserved.
```
Comment (saadrahim) [inline HIP-Basic/assembly_to_executable/Makefile]: Copyright

File: HIP-Basic/assembly_to_executable/assembly_to_executable_vs2019.vcxproj.filters [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (saadrahim) [inline HIP-Basic/assembly_to_executable/assembly_to_executable_vs2019.vcxproj.filters:1]: For file names and folders, are using -, _, or CamelCase or camelCase? Let's add it in a style guide and stick with it.

File: HIP-Basic/runtime_compilation/README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
+# HIP-Basic Runtime Compilation Example
+
+## Description
+Runtime compilation allows compiling code to machine code while a program is being executed rather than before, as happens with static compilation.
```
Comment (saadrahim) [inline HIP-Basic/runtime_compilation/README.md]: rather than at compile, need to rephrase this

File: HIP-Basic/runtime_compilation/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+HIP supports runtime compilation (hipRTC), allowing to compile certain fragments of code in runtime and thus permitting runtime optimizations in them.
+Therefore, the usage of hipRTC provides the possibility of obtaining optimizations and performance improvements compared with other APIs with regular offline static compilation.
+
+This example showcases how to make use of hipRTC to compile in runtime a kernel and launch it on a device. This kernel is a simple SAXPY, i.e. a single-precision operation $y_i=ax_i+y_i$.
+
+### Application flow 
+1. A number of variables are declared and defined to configure the program which will be compiled in runtime.
+2. The program is created using the above variables as parameters, along with the SAXPY kernel in string form.
+3. The properties of the first device (GPU) available are consulted to set the device architecture as (the only) compile option.
+4. The program is compiled using the previously mentioned compile options.
```
Comment (saadrahim) [inline HIP-Basic/runtime_compilation/README.md:14]: Let's have a flow diagram

File: HIP-Basic/warp_shuffle/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
+# HIP-Basic Warp Shuffle Example
+
+## Description
+Kernel code is SIMT-executed by groups of threads within the same block, known as _warps_ (NVIDIA) or _wavefronts_ (AMD). Each block is divided into as many warps as the block's size allows, if the block size is less than the warp size then part of the warp just stays idle (as happens in this example).
```
Comment (saadrahim) [inline HIP-Basic/warp_shuffle/README.md]: SIMT or SIMD?, please define acronymn

File: HIP-Basic/warp_shuffle/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# HIP-Basic Warp Shuffle Example
+
+## Description
+Kernel code is SIMT-executed by groups of threads within the same block, known as _warps_ (NVIDIA) or _wavefronts_ (AMD). Each block is divided into as many warps as the block's size allows, if the block size is less than the warp size then part of the warp just stays idle (as happens in this example). 
+In NVIDIA GPUs the warp size is 32 threads, while in AMDs it is 64 threads for architectures previous to RDNA1. After RDNA1 all architectures support both 32 and 64 warp sizes.
```
Comment (saadrahim) [inline HIP-Basic/warp_shuffle/README.md]: We have guidance for referring to RDNA, we need make sure this conforms to it. I believe there needs to be TM the first time.

**Review by saadrahim — CHANGES_REQUESTED** [review]
Please review the readme language. A few diagrams will be great. A warp shuffle diagram and hipRTC process are two that I want.

File: HIP-Basic/assembly_to_executable/Makefile (hunk 1 above)
Comment (Snektron) [reply HIP-Basic/assembly_to_executable/Makefile]: Updated to the current year.

File: HIP-Basic/assembly_to_executable/assembly_to_executable_vs2019.vcxproj.filters (hunk 2 above)
Comment (Snektron) [reply HIP-Basic/assembly_to_executable/assembly_to_executable_vs2019.vcxproj.filters:1]: Directories use pascal case in general, with the exception of those in Libraries, which should follow the same casing style as the library they represent. Files and directories that represent an example use snake case. We've added this to the guide in Docs/, and re-named the directories that did not follow it accordingly.

File: HIP-Basic/runtime_compilation/README.md (hunk 3 above)
Comment (Snektron) [reply HIP-Basic/runtime_compilation/README.md]: Its rephrased a bit.

File: HIP-Basic/runtime_compilation/README.md (hunk 4 above)
Comment (Snektron) [reply HIP-Basic/runtime_compilation/README.md:14]: Added a flow diagram in HIP-Basic/runtime_compilation/hiprtc.svg. You can see a preview [here](https://github.com/amd/rocm-examples/tree/53bad130d49b27421e32478b1194a8723c0d2edc/HIP-Basic/runtime_compilation).

File: HIP-Basic/warp_shuffle/README.md (hunk 5 above)
Comment (Snektron) [reply HIP-Basic/warp_shuffle/README.md]: Rephrased this and the next paragraph a bit.

File: HIP-Basic/warp_shuffle/README.md (hunk 6 above)
Comment (Snektron) [reply HIP-Basic/warp_shuffle/README.md]: Added

File: HIP-Basic/warp_shuffle/README.md (hunk 5 above)
Comment (Snektron) [reply HIP-Basic/warp_shuffle/README.md]: See also the diagrams now added to this readme [here](https://github.com/amd/rocm-examples/tree/53bad130d49b27421e32478b1194a8723c0d2edc/HIP-Basic/warp_shuffle).

File: HIP-Basic/runtime_compilation/README.md (hunk 4 above)
Comment (saadrahim) [reply HIP-Basic/runtime_compilation/README.md:14]: nice

**Review by saadrahim — APPROVED** [review]

**Pull request merged**